contracts/tests/bridge/test_measure_gas_home_bridge.py

```diff
@@ -0,0 +1,120 @@
+    """the number of confirmations required"""
+    return (number_of_validators * 50 + 99) // 100
+
+
+@pytest.fixture(scope="session")
+def all_proxy_validators(chain):
+    """list of 123 accounts that can be used as validators"""
+    account_0 = chain.get_accounts()[0]
+    validators = []
+    for i in range(maximal_number_of_validators):
+        account_num = 20000000 + i  # do not
+        new_account = chain.add_account(f"0x{account_num:064}")
+        chain.send_transaction(
+            {
+                "from": account_0,
+                "to": new_account,
+                "gas": 100000,
+                "value": 1_000_000_000,
+            }
+        )
+        validators.append(new_account)
+
+    return validators
+
+
+@pytest.fixture()
+def proxy_validators(all_proxy_validators, number_of_validators):
+    """The validators used in the proxy contract. This replaces the
+    fixture with the same name from conftest.py"""
+    return all_proxy_validators[:number_of_validators]
+
+
+@pytest.fixture
+def confirm_nth(home_bridge_contract, proxy_validators, web3):
+    """confirm a transfer by the nth validator"""
+
+    class ConfirmNth:
+        transfer_hash = "0x" + b"     transfer-hash              ".hex()
+        tx_hash = "0x" + b"     tx-hash                    ".hex()
+        amount = 20000
+        recipient = "0xFCB047cCD297048b6F31fbb2fef14001FefFa0f3"
+
+        def __call__(self, n, fail_ok=False):
+            validator = proxy_validators[n]
+            transact_args = {"from": validator}
+            if fail_ok:
+                # we pass in the gas here in order to not make it not
+                # raise while estimating gas
+                transact_args["gas"] = 2_000_000
+
+            tx_hash = home_bridge_contract.functions.confirmTransfer(
+                transferHash=self.transfer_hash,
+                transactionHash=self.tx_hash,
+                amount=self.amount,
+                recipient=self.recipient,
+            ).transact(transact_args)
+            tx_receipt = web3.eth.getTransactionReceipt(tx_hash)
+            gas = tx_receipt.gasUsed
+            print(
+                f"validator {n+1} {validator} confirmed, gas: {gas}, status: {tx_receipt.status}"
+            )
+            return gas
+
+    return ConfirmNth()
+
+
+def test_gas_cost_complete_transfer(
+    home_bridge_contract,
+    proxy_validators,
+    confirm_nth,
+    web3,
+    number_of_validators,
+    required_confirmations,
+):
+    """This walks through a complete Transfer on the home bridge and
+    checks gas usage limits
+    """
+    print(
+        f"\n=====> {number_of_validators} validators, {required_confirmations} confirmations required"
+    )
+    get_transfer_completed_events = home_bridge_contract.events.TransferCompleted.createFilter(
+        fromBlock=web3.eth.blockNumber
+    ).get_all_entries
+
+    for i in range(required_confirmations - 1):
+        gas = confirm_nth(i)
+        assert gas < 120_000
```

Comment:
Wouldn't it make sense to print the gas usage (maybe median/average/max/min for all of them)? Or how did you get the numbers out?

Reply:
It's printed in `confirm_nth`. Output looks like the following. Printing `min/max/median/average` doesn't provide much additional information.

```
..ocol/blockchain/contracts/tests/bridge (git)-[measure-home-bridge-gas-usage] % ./test_measure_gas_home_bridge.py 
======================================================= test session starts ========================================================
platform linux -- Python 3.6.8, pytest-4.6.2, py-1.8.0, pluggy-0.12.0
solc: /home/bbralf/bin/solc
solidity version: 0.5.8+commit.23d335f2.Linux.g++

rootdir: /home/bbralf/trustlines/protocol/blockchain/contracts, inifile: pytest.ini
plugins: contract-deploy-tools-0.4.3, ethereum-0.1.3a6
collected 2 items                                                                                                                  

test_measure_gas_home_bridge.py 
=====> 123 validators, 62 confirmations required
validator 1 0xb1C35866AEba18de80b8A60226EA47990F7D2208 confirmed, gas: 117276, status: 1
validator 2 0x207ff8804C1B9f7D360De4909b7840b7812C5188 confirmed, gas: 87276, status: 1
validator 3 0x77F711dbFba758b9587d2b0d3D86E1DD81274ac7 confirmed, gas: 87276, status: 1
validator 4 0xa120345109eea4A153D05FE23864B91CB00c5dA8 confirmed, gas: 87276, status: 1
validator 5 0xe287a58C6eE5Dee6c506921d7DDEfa32Bc413bdE confirmed, gas: 87276, status: 1
validator 6 0x810482D0D30e5fACc24971143D5c9D2665D0E66A confirmed, gas: 87276, status: 1
validator 7 0xAF3e1dd74233f05E967C3935bD623331743cD6e4 confirmed, gas: 87276, status: 1
validator 8 0x5166cb1dc078856d82404b7B4a2F31fA76f5dAcc confirmed, gas: 87276, status: 1
validator 9 0x2e4d29523Aa3414C861792F0B49dc9D144653B75 confirmed, gas: 87276, status: 1
validator 10 0x768171827B61d2a544D9d9F53aC8608179528b9d confirmed, gas: 87276, status: 1
validator 11 0x6d491BBBc2A6260FFcF990141E28d59516Ec6a69 confirmed, gas: 87276, status: 1
validator 12 0xCF59d400e51dEaF55626B931cA181a936bf16B49 confirmed, gas: 87276, status: 1
validator 13 0x293c7056FCA01B928ca4a46c965a7244DAD8a36B confirmed, gas: 87276, status: 1
validator 14 0xE3476Be77C493820e52AA621cB2770201b47D6ae confirmed, gas: 87276, status: 1
validator 15 0x13c533bFcf0a9d6aDF94475D4B28a18D865403D6 confirmed, gas: 87276, status: 1
validator 16 0x32e0225A14a373D400d5e936Fbb8F49569a03565 confirmed, gas: 87276, status: 1
validator 17 0x00c8F76eF86D69e8A88e37912b120d9C0199e3C2 confirmed, gas: 87276, status: 1
validator 18 0xf2b551ff0f1020E1CCFd71d4f04a54C31AF3eC53 confirmed, gas: 87276, status: 1
validator 19 0x78DD98418B56e9A3866227bF4415467be96Aa1ca confirmed, gas: 87276, status: 1
validator 20 0xb7AB41D88c4F262985bB9C10A515803b4f1b8196 confirmed, gas: 87276, status: 1
validator 21 0x26182B9545Fea7813B1Cef704c4F6bE2dc0F3a29 confirmed, gas: 87276, status: 1
validator 22 0x12180459EeBD3DE070565Aff99Ee735d5691931F confirmed, gas: 87276, status: 1
validator 23 0xE7C61cb32389b25f024c5b79Df62B6BB73BAafBd confirmed, gas: 87276, status: 1
validator 24 0x9061637229c4D86Ecb736A961210749453B8De99 confirmed, gas: 87276, status: 1
validator 25 0xe4c42f8d44Da73A4bFFd6d45D7D7245f5F8DEa98 confirmed, gas: 87276, status: 1
validator 26 0xB23E10e561176472D93cd9fC44e1a0EdD1bBa4C7 confirmed, gas: 87276, status: 1
validator 27 0xe18a7e038C9da6bf3A2a60B499eD653CB8465397 confirmed, gas: 87276, status: 1
validator 28 0xf5a2Ed5b0280C6aeF831AdC6c6248cb6A4d360a9 confirmed, gas: 87276, status: 1
validator 29 0x86bA6f3F500A02c686f68712BF735bDD51277CF4 confirmed, gas: 87276, status: 1
validator 30 0xAD185C6E176A2aBE70A89673F8E77650697484EC confirmed, gas: 87276, status: 1
validator 31 0xcD8AFAFFDD4e903D321fa2F72fCEbb78D8e61764 confirmed, gas: 87276, status: 1
validator 32 0x85375D989626c0510761EdB10F874B3D62637083 confirmed, gas: 87276, status: 1
validator 33 0x561b614FDEda59329620054de0F52981Cd8Cf93B confirmed, gas: 87276, status: 1
validator 34 0xBbA85b0CCD7C0CaE4794F5AE479D2525F8da0B3e confirmed, gas: 87276, status: 1
validator 35 0xe779547B5d82fd87ed6f13924fEA019f2091E85b confirmed, gas: 87276, status: 1
validator 36 0xAd0DF000900C804aC1Ca3Df716e291561A32CF7A confirmed, gas: 87276, status: 1
validator 37 0x6e1fA35CE54F767FBf49137Cd765ad1E523d81FA confirmed, gas: 87276, status: 1
validator 38 0xfdd360F852a7FF873b8C66F1dB48aa791CeA2513 confirmed, gas: 87276, status: 1
validator 39 0x47DA5a61108F70eBED9af8da5BcC9401eE4De000 confirmed, gas: 87276, status: 1
validator 40 0x1B83391Fe8237B6d98690388ae93DF7EE6Fdb0ce confirmed, gas: 87276, status: 1
validator 41 0x73fEaCB10CA72659817799d44e9B6Fe919e3b3C9 confirmed, gas: 87276, status: 1
validator 42 0xA0a8a00d434D1307AA3a6380B694844Ea0267D2A confirmed, gas: 87276, status: 1
validator 43 0xbDAc9E1e1cD516E6043B9AB0eF9ac705F47f5630 confirmed, gas: 87276, status: 1
validator 44 0xD006C3ea85F20BAbEF06ba0460e99Bd9EC80c744 confirmed, gas: 87276, status: 1
validator 45 0x39a2Bd8b722dd4adbdCc19359bc94509D331389e confirmed, gas: 87276, status: 1
validator 46 0x2948b4aC2369dbFf9d552454037821c1C63A951D confirmed, gas: 87276, status: 1
validator 47 0x29309629471100a81801f29e9b7989729d8aF48b confirmed, gas: 87276, status: 1
validator 48 0x6801c010E2B560f888BCF73e3a56Bb6Ad9d52032 confirmed, gas: 87276, status: 1
validator 49 0x5110c059DF48D12a0Df6CDA3FEdDA1b418d42C08 confirmed, gas: 87276, status: 1
validator 50 0xD01C8C25E104F8FaF71513A04DBF5A7c0F5c50DE confirmed, gas: 87276, status: 1
validator 51 0x8BeC12958ba5C7479027AF0a1daAf1dDfCD67B58 confirmed, gas: 87276, status: 1
validator 52 0xcEB9435A45E4F8340fD70285dFC4e983F2ea62C2 confirmed, gas: 87276, status: 1
validator 53 0x6E4828Bb4420E4324DdAbFcD99d32d2773BF3dfa confirmed, gas: 87276, status: 1
validator 54 0x9916Cac3488b0E5661EF81Df8e71F32b00f0A02E confirmed, gas: 87276, status: 1
validator 55 0x4e90D6487968e2b943F07ce189FAdB11BF73627e confirmed, gas: 87276, status: 1
validator 56 0xA673523BC874f4F5DDeBaAD883066278Eb7B1Abf confirmed, gas: 87276, status: 1
validator 57 0x88735A28249255AE4FF2041B51d43A349302f224 confirmed, gas: 87276, status: 1
validator 58 0x88f9663B64F8e35f3EBa9d4daD587905C06f9fa4 confirmed, gas: 87276, status: 1
validator 59 0x512A4072490b82e680ab1831D3262A133892e1B8 confirmed, gas: 87276, status: 1
validator 60 0xF231C7dA410a39713DF3548245cd7922dfef9e34 confirmed, gas: 87276, status: 1
validator 61 0xA9e74d49a97002D87FD2214891128707926BA6e2 confirmed, gas: 87276, status: 1
Completing the transfer
validator 62 0xc5eDf86ac11eB1B68371d2Cc911737c996C18F5f confirmed, gas: 320777, status: 1
Confirm after complete
validator 63 0xe947EF8699Df5B2Df1e79915e4da3474E8C8576b confirmed, gas: 30714, status: 0

.
=====> 50 validators, 25 confirmations required
validator 1 0xb1C35866AEba18de80b8A60226EA47990F7D2208 confirmed, gas: 117276, status: 1
validator 2 0x207ff8804C1B9f7D360De4909b7840b7812C5188 confirmed, gas: 87276, status: 1
validator 3 0x77F711dbFba758b9587d2b0d3D86E1DD81274ac7 confirmed, gas: 87276, status: 1
validator 4 0xa120345109eea4A153D05FE23864B91CB00c5dA8 confirmed, gas: 87276, status: 1
validator 5 0xe287a58C6eE5Dee6c506921d7DDEfa32Bc413bdE confirmed, gas: 87276, status: 1
validator 6 0x810482D0D30e5fACc24971143D5c9D2665D0E66A confirmed, gas: 87276, status: 1
validator 7 0xAF3e1dd74233f05E967C3935bD623331743cD6e4 confirmed, gas: 87276, status: 1
validator 8 0x5166cb1dc078856d82404b7B4a2F31fA76f5dAcc confirmed, gas: 87276, status: 1
validator 9 0x2e4d29523Aa3414C861792F0B49dc9D144653B75 confirmed, gas: 87276, status: 1
validator 10 0x768171827B61d2a544D9d9F53aC8608179528b9d confirmed, gas: 87276, status: 1
validator 11 0x6d491BBBc2A6260FFcF990141E28d59516Ec6a69 confirmed, gas: 87276, status: 1
validator 12 0xCF59d400e51dEaF55626B931cA181a936bf16B49 confirmed, gas: 87276, status: 1
validator 13 0x293c7056FCA01B928ca4a46c965a7244DAD8a36B confirmed, gas: 87276, status: 1
validator 14 0xE3476Be77C493820e52AA621cB2770201b47D6ae confirmed, gas: 87276, status: 1
validator 15 0x13c533bFcf0a9d6aDF94475D4B28a18D865403D6 confirmed, gas: 87276, status: 1
validator 16 0x32e0225A14a373D400d5e936Fbb8F49569a03565 confirmed, gas: 87276, status: 1
validator 17 0x00c8F76eF86D69e8A88e37912b120d9C0199e3C2 confirmed, gas: 87276, status: 1
validator 18 0xf2b551ff0f1020E1CCFd71d4f04a54C31AF3eC53 confirmed, gas: 87276, status: 1
validator 19 0x78DD98418B56e9A3866227bF4415467be96Aa1ca confirmed, gas: 87276, status: 1
validator 20 0xb7AB41D88c4F262985bB9C10A515803b4f1b8196 confirmed, gas: 87276, status: 1
validator 21 0x26182B9545Fea7813B1Cef704c4F6bE2dc0F3a29 confirmed, gas: 87276, status: 1
validator 22 0x12180459EeBD3DE070565Aff99Ee735d5691931F confirmed, gas: 87276, status: 1
validator 23 0xE7C61cb32389b25f024c5b79Df62B6BB73BAafBd confirmed, gas: 87276, status: 1
validator 24 0x9061637229c4D86Ecb736A961210749453B8De99 confirmed, gas: 87276, status: 1
Completing the transfer
validator 25 0xe4c42f8d44Da73A4bFFd6d45D7D7245f5F8DEa98 confirmed, gas: 202636, status: 1
Confirm after complete
validator 26 0xB23E10e561176472D93cd9fC44e1a0EdD1bBa4C7 confirmed, gas: 30714, status: 0

.

==================================================== 2 passed in 41.45 seconds =====================================================
.
```